tests/cases/unittests/session.ts

```diff
@@ -0,0 +1,459 @@
+
+				expect(() => session.addProtocolHandler(command, (req) => resp))
+				.to.throw(`Protocol handler already exists for command "${command}"`);
+			});
+		});
+		
+		describe('event', () => {
+			it('can format event responses and send them', () => {
+				let evt = 'notify-test',
+				info = {
+					test: true
+				};
+
+				session.event(info, evt);
+
+				expect(lastSent).to.deep.equal({
+					type: 'event',
+					seq: 0,
+					event: evt,
+					body: info
+				});
+			});
+		});
+
+		describe('output', () => {
+			it('can format command responses and send them', () => {
+				let body = {
+					block: {
+						key: 'value'
+					}
+				},
+				command = 'test';
+
+				session.output(body, command);
+
+				expect(lastSent).to.deep.equal({
+					seq: 0,
+					request_seq: 0,
+					type: 'response',
+					command,
+					body: body,
+					success: true
+				});
+			});
+		});
+	});
+
+	describe('how Session is extendable via subclassing', () => {
+		let TestSession = class extends Session {
+			lastSent: protocol.Message;
+			customHandler: string = 'testhandler';
+			constructor(){
+				super(mockHost, Buffer.byteLength, process.hrtime, mockLogger);
+				this.addProtocolHandler(this.customHandler, () => {
+					return {response: undefined, responseRequired: true};
+				});
+			}
+			send(msg: protocol.Message) {
+				this.lastSent = msg;
+			}
+		};
+
+		it('can override methods such as send', () => {
+			let session = new TestSession(),
+			body = {
+				block: {
+					key: 'value'
+				}
+			},
+			command = 'test';
+
+			session.output(body, command);
+
+			expect(session.lastSent).to.deep.equal({
+				seq: 0,
+				request_seq: 0,
+				type: 'response',
+				command,
+				body: body,
+				success: true
+			});
+		});
+		it('can add and respond to new protocol handlers', () => {
+			let session = new TestSession();
+
+			expect(session.executeCommand({
+				seq: 0,
+				type: 'command',
+				command: session.customHandler
+			})).to.deep.equal({
+				response: undefined,
+				responseRequired: true
+			});
+		});
+		it('has access to the project service', () => {
+			let ServiceSession = class extends TestSession {
+				constructor() {
+					super();
+					assert(this.projectService);
+					expect(this.projectService).to.be.instanceOf(ProjectService);
+				}
+			};
+			new ServiceSession();
+		});
+	});
+
+	describe('an example of using the Session API to create an in-process server', () => {
+		let inProcHost: ServerHost = {
+			args: [],
+			newLine: '\n',
+			useCaseSensitiveFileNames: true,
+			write: (s) => lastWrittenToHost = s,
+			readFile: () => void 0,
+			writeFile: () => void 0,
+			resolvePath: () => void 0,
+			fileExists: () => false,
+			directoryExists: () => false,
+			createDirectory: () => void 0,
+			getExecutingFilePath: () => void 0,
+			getCurrentDirectory: () => void 0,
+			readDirectory: () => void 0,
+			exit: () => void 0
+		},
+		InProcSession = class extends Session {

```

Comment:
`class declarations are only supported directly inside a module or as a top-level declaration` at least that's the error on a TS playground (and likely in VS Code (which is what I use most of the time) as well)  - If this isn't true for ToT TS then it's my fault for trusting my editor too much. (I did explicitly ignore its errors about class expressions not being allowed since I had read over the PR adding them.) 

I have yet to set up Code in such a way where is can track the latest TS while I'm editing the latest TS, so it may be my fault for thinking that. Actual class declarations are way more idiomatic here if they're actually allowed in that context.